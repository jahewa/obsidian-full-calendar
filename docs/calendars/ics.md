# Remote calendars in .ics format

Add any calendar that you have a link to in .ics format to Obsidian. This includes public calendars like [this one of US holidays](https://www.officeholidays.com/subscribe/usa), but also includes [private Google Calendars](https://support.google.com/calendar/answer/37648?hl=en#zippy=%2Csync-your-google-calendar-view-edit%2Cget-your-calendar-view-only%2Csecret-address) and [public Apple Calendars](https://support.apple.com/guide/icloud/share-a-calendar-mm6b1a9479/icloud). The walkthrough below shows where to find a Google Calendar's private .ics link, and how to add it to Obsidian, but any URL will work just as well.

Calendars are re-fetched automatically from their source at most every five minutes. If you would like to revalidate remote calendars directly, you can run the command `Full Calendar: Revalidate remote calendars`.

![](../assets/sync-setup-ics.gif)
