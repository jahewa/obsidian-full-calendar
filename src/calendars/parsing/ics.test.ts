import { getEventsFromICS } from "./ics";

describe("ics tests", () => {
    it("parses all day event", () => {
        const ics = `BEGIN:VCALENDAR
PRODID:blah
X-WR-CALNAME:Test calendar
X-WR-TIMEZONE:Etc/UTC
VERSION:2.0
CALSCALE:GREGORIAN
X-PUBLISHED-TTL:PT5M
METHOD:PUBLISH
BEGIN:VEVENT
UID:7389432083-0-40713-74006
SEQUENCE:1
CLASS:PUBLIC
CREATED:20200101T000000Z
GEO:40.7128;-74.006
DTSTAMP:20230226T143136Z
DTSTART;VALUE=DATE:20230226
DESCRIPTION:Description!
LOCATION:New york city
URL:https://www.example.com
STATUS:CONFIRMED
SUMMARY:EVENT TITLE
TRANSP:TRANSPARENT
END:VEVENT
END:VCALENDAR`;
        const events = getEventsFromICS(ics);
        expect(events).toMatchSnapshot(ics);
    });

    it("parses gcal ics file", () => {
        const ics = `BEGIN:VCALENDAR
PRODID:-//Google Inc//Google Calendar 70.9054//EN
VERSION:2.0
CALSCALE:GREGORIAN
METHOD:PUBLISH
X-WR-CALNAME:Obsidian Test Calendar
X-WR-TIMEZONE:America/New_York
BEGIN:VTIMEZONE
TZID:America/New_York
X-LIC-LOCATION:America/New_York
BEGIN:DAYLIGHT
TZOFFSETFROM:-0500
TZOFFSETTO:-0400
TZNAME:EDT
DTSTART:19700308T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=2SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:-0400
TZOFFSETTO:-0500
TZNAME:EST
DTSTART:19701101T020000
RRULE:FREQ=YEARLY;BYMONTH=11;BYDAY=1SU
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
DTSTART;VALUE=DATE:20220302
DTEND;VALUE=DATE:20220303
DTSTAMP:20230302T233513Z
UID:5r09pnnlktaqivstai5vlbqb1h@google.com
CREATED:20220226T211158Z
DESCRIPTION:
LAST-MODIFIED:20220226T214634Z
LOCATION:
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:All day event
TRANSP:TRANSPARENT
END:VEVENT
BEGIN:VEVENT
DTSTART;TZID=America/New_York:20220301T110000
DTEND;TZID=America/New_York:20220301T123000
RRULE:FREQ=WEEKLY;WKST=SU;BYDAY=TH,TU
DTSTAMP:20230302T233513Z
UID:5tt2avr2th0h65homv3b6jeqof@google.com
CREATED:20220226T211144Z
DESCRIPTION:
LAST-MODIFIED:20220226T214627Z
LOCATION:
SEQUENCE:1
STATUS:CONFIRMED
SUMMARY:Recurring event
TRANSP:OPAQUE
END:VEVENT
BEGIN:VEVENT
DTSTART:20220228T164500Z
DTEND:20220228T194500Z
DTSTAMP:20230302T233513Z
UID:40mdbe6fvc1rmd60n6r0c3go7e@google.com
X-GOOGLE-CONFERENCE:https://meet.google.com/riu-josb-pdb
CREATED:20220226T210517Z
DESCRIPTION:This is an example <i>event.</i>\n\nJoin with Google Meet: http
    s://meet.google.com/riu-josb-pdb\nOr dial: (US) +1 609-726-6186 PIN: 156393
    865#\nMore phone numbers: https://tel.meet/riu-josb-pdb?pin=1416269198709&h
    s=7\n\nLearn more about Meet at: https://support.google.com/a/users/answer/
    9282720
LAST-MODIFIED:20220226T214608Z
LOCATION:
SEQUENCE:1
STATUS:CONFIRMED
SUMMARY:Hello\, iCal!
TRANSP:OPAQUE
END:VEVENT
BEGIN:VEVENT
DTSTART:20220219T190000Z
DTEND:20220219T230000Z
DTSTAMP:20230302T233513Z
UID:44hekcaaf0or7547vhqa772mqj@google.com
CREATED:20220220T002201Z
DESCRIPTION:
LAST-MODIFIED:20220220T002201Z
LOCATION:
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:Work on GCal Sync
TRANSP:OPAQUE
END:VEVENT
BEGIN:VEVENT
DTSTART;VALUE=DATE:20220216
DTEND;VALUE=DATE:20220217
DTSTAMP:20230302T233513Z
UID:7ooluqb717vabebvc9gkc38c9l@google.com
CREATED:20220220T002146Z
DESCRIPTION:
LAST-MODIFIED:20220220T002146Z
LOCATION:
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:Announce Beta
TRANSP:TRANSPARENT
END:VEVENT
END:VCALENDAR
        `;
        const events = getEventsFromICS(ics);
        expect(events).toMatchSnapshot(ics);
    });
});
