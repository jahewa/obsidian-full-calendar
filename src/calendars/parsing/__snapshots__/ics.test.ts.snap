// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ics tests parses all day event: BEGIN:VCALENDAR
PRODID:blah
X-WR-CALNAME:Test calendar
X-WR-TIMEZONE:Etc/UTC
VERSION:2.0
CALSCALE:GREGORIAN
X-PUBLISHED-TTL:PT5M
METHOD:PUBLISH
BEGIN:VEVENT
UID:7389432083-0-40713-74006
SEQUENCE:1
CLASS:PUBLIC
CREATED:20200101T000000Z
GEO:40.7128;-74.006
DTSTAMP:20230226T143136Z
DTSTART;VALUE=DATE:20230226
DESCRIPTION:Description!
LOCATION:New york city
URL:https://www.example.com
STATUS:CONFIRMED
SUMMARY:EVENT TITLE
TRANSP:TRANSPARENT
END:VEVENT
END:VCALENDAR 1`] = `
[
  {
    "allDay": true,
    "date": "2023-02-26",
    "endDate": null,
    "id": "ics::7389432083-0-40713-74006::2023-02-26::single",
    "title": "EVENT TITLE",
    "type": "single",
  },
]
`;

exports[`ics tests parses gcal ics file: BEGIN:VCALENDAR
PRODID:-//Google Inc//Google Calendar 70.9054//EN
VERSION:2.0
CALSCALE:GREGORIAN
METHOD:PUBLISH
X-WR-CALNAME:Obsidian Test Calendar
X-WR-TIMEZONE:America/New_York
BEGIN:VTIMEZONE
TZID:America/New_York
X-LIC-LOCATION:America/New_York
BEGIN:DAYLIGHT
TZOFFSETFROM:-0500
TZOFFSETTO:-0400
TZNAME:EDT
DTSTART:19700308T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=2SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:-0400
TZOFFSETTO:-0500
TZNAME:EST
DTSTART:19701101T020000
RRULE:FREQ=YEARLY;BYMONTH=11;BYDAY=1SU
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
DTSTART;VALUE=DATE:20220302
DTEND;VALUE=DATE:20220303
DTSTAMP:20230302T233513Z
UID:5r09pnnlktaqivstai5vlbqb1h@google.com
CREATED:20220226T211158Z
DESCRIPTION:
LAST-MODIFIED:20220226T214634Z
LOCATION:
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:All day event
TRANSP:TRANSPARENT
END:VEVENT
BEGIN:VEVENT
DTSTART;TZID=America/New_York:20220301T110000
DTEND;TZID=America/New_York:20220301T123000
RRULE:FREQ=WEEKLY;WKST=SU;BYDAY=TH,TU
DTSTAMP:20230302T233513Z
UID:5tt2avr2th0h65homv3b6jeqof@google.com
CREATED:20220226T211144Z
DESCRIPTION:
LAST-MODIFIED:20220226T214627Z
LOCATION:
SEQUENCE:1
STATUS:CONFIRMED
SUMMARY:Recurring event
TRANSP:OPAQUE
END:VEVENT
BEGIN:VEVENT
DTSTART:20220228T164500Z
DTEND:20220228T194500Z
DTSTAMP:20230302T233513Z
UID:40mdbe6fvc1rmd60n6r0c3go7e@google.com
X-GOOGLE-CONFERENCE:https://meet.google.com/riu-josb-pdb
CREATED:20220226T210517Z
DESCRIPTION:This is an example <i>event.</i>

Join with Google Meet: http
    s://meet.google.com/riu-josb-pdb
Or dial: (US) +1 609-726-6186 PIN: 156393
    865#
More phone numbers: https://tel.meet/riu-josb-pdb?pin=1416269198709&h
    s=7

Learn more about Meet at: https://support.google.com/a/users/answer/
    9282720
LAST-MODIFIED:20220226T214608Z
LOCATION:
SEQUENCE:1
STATUS:CONFIRMED
SUMMARY:Hello, iCal!
TRANSP:OPAQUE
END:VEVENT
BEGIN:VEVENT
DTSTART:20220219T190000Z
DTEND:20220219T230000Z
DTSTAMP:20230302T233513Z
UID:44hekcaaf0or7547vhqa772mqj@google.com
CREATED:20220220T002201Z
DESCRIPTION:
LAST-MODIFIED:20220220T002201Z
LOCATION:
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:Work on GCal Sync
TRANSP:OPAQUE
END:VEVENT
BEGIN:VEVENT
DTSTART;VALUE=DATE:20220216
DTEND;VALUE=DATE:20220217
DTSTAMP:20230302T233513Z
UID:7ooluqb717vabebvc9gkc38c9l@google.com
CREATED:20220220T002146Z
DESCRIPTION:
LAST-MODIFIED:20220220T002146Z
LOCATION:
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:Announce Beta
TRANSP:TRANSPARENT
END:VEVENT
END:VCALENDAR
         1`] = `
[
  {
    "allDay": true,
    "date": "2022-03-02",
    "endDate": "2022-03-03",
    "id": "ics::5r09pnnlktaqivstai5vlbqb1h@google.com::2022-03-02::single",
    "title": "All day event",
    "type": "single",
  },
  {
    "allDay": false,
    "endTime": "12:30",
    "id": "ics::5tt2avr2th0h65homv3b6jeqof@google.com::2022-03-01::recurring",
    "rrule": "RRULE:FREQ=WEEKLY;BYDAY=TH,TU;WKST=SU",
    "skipDates": [],
    "startDate": "2022-03-01",
    "startTime": "11:00",
    "title": "Recurring event",
    "type": "rrule",
  },
  {
    "allDay": false,
    "date": "2022-02-28",
    "endDate": null,
    "endTime": "19:45",
    "id": "ics::40mdbe6fvc1rmd60n6r0c3go7e@google.com::2022-02-28::single",
    "startTime": "16:45",
    "title": "Hello, iCal!",
    "type": "single",
  },
  {
    "allDay": false,
    "date": "2022-02-19",
    "endDate": null,
    "endTime": "23:00",
    "id": "ics::44hekcaaf0or7547vhqa772mqj@google.com::2022-02-19::single",
    "startTime": "19:00",
    "title": "Work on GCal Sync",
    "type": "single",
  },
  {
    "allDay": true,
    "date": "2022-02-16",
    "endDate": "2022-02-17",
    "id": "ics::7ooluqb717vabebvc9gkc38c9l@google.com::2022-02-16::single",
    "title": "Announce Beta",
    "type": "single",
  },
]
`;
